# Exercise 1

### Step 1. Go to [Kaggle]( https://www.kaggle.com/openfoodfacts/world-food-facts)

### Step 2. Download the dataset to your computer and unzip it.

El dataset que s'ha descarregat conté informació de diferents aliments de diversos països arreu del món. Aquesta informació es basa principalment en la composició nutricional i ingredients que poden causar al·lèrgies de cada un dels productes.

### Step 3. Use the csv file and assign it to a dataframe called food

Per tal de realitzar l'exercici, carregam la llibreria <span style="color:#DF7401">**pandas**</span>.

In [1]:
import pandas as pd

Ara, declaram el dataframe _food_, amb la informació del conjunt de dades descarregat a **Step 2**.

In [2]:
food = pd.read_csv('data/en.openfoodfacts.org.products.tsv',sep='\t')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2698: DtypeWarning: Columns (0,3,5,19,20,24,25,26,27,28,36,37,38,39,48) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


### Step 4. See the first 5 entries

Amb la funció _head()_ podem visualitzar les primeres observacions de _food_. En aquest cas, li especificam que volem les 5 primeres files.

In [3]:
food.head(5)

code                                                url  \
0   3087  http://world-en.openfoodfacts.org/product/0000...   
1   4530  http://world-en.openfoodfacts.org/product/0000...   
2   4559  http://world-en.openfoodfacts.org/product/0000...   
3  16087  http://world-en.openfoodfacts.org/product/0000...   
4  16094  http://world-en.openfoodfacts.org/product/0000...   

                      creator   created_t      created_datetime  \
0  openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
1             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
2             usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
3             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
4             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

  last_modified_t last_modified_datetime                    product_name  \
0      1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
1      1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
2      1489069957   2017-03-09T14:32:37Z                         Peanuts   
3      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
4      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

  generic_name quantity         ...         fruits-vegetables-nuts_100g  \
0          NaN      1kg         ...                                 NaN   
1          NaN      NaN         ...                                 NaN   
2          NaN      NaN         ...                                 NaN   
3          NaN      NaN         ...                                 NaN   
4          NaN      NaN         ...                                 NaN   

  fruits-vegetables-nuts-estimate_100g collagen-meat-protein-ratio_100g  \
0                                  NaN                              NaN   
1                                  NaN                              NaN   
2                                  NaN                              NaN   
3                                  NaN                              NaN   
4                                  NaN                              NaN   

  cocoa_100g chlorophyl_100g carbon-footprint_100g nutrition-score-fr_100g  \
0        NaN             NaN                   NaN                     NaN   
1        NaN             NaN                   NaN                    14.0   
2        NaN             NaN                   NaN                     0.0   
3        NaN             NaN                   NaN                    12.0   
4        NaN             NaN                   NaN                     NaN   

  nutrition-score-uk_100g glycemic-index_100g water-hardness_100g  
0                     NaN                 NaN                 NaN  
1                    14.0                 NaN                 NaN  
2                     0.0                 NaN                 NaN  
3                    12.0                 NaN                 NaN  
4                     NaN                 NaN                 NaN  

[5 rows x 163 columns]

Si ens fixam en les darreres columnes, sembla que falta informació sobre alguns dels productes, ja que hi ha molts NaN, almenys en les 5 primeres observacions.


### Step 5. What is the number of observations in the dataset?

La funció _len()_ aplicada a un dataframe retorna el nombre d'observacions. Així, l'aplicam a _food_ per conèixer de quants productes s'han recollit les dades.

In [4]:
len(food)

356027

### Step 6. What is the number of columns in the dataset?

D'altra banda, la següent instrucció torna el nombre de columnes del dataframe i, per tant, el nombre d'informació que es pretén recollir de cada producte.

In [5]:
len(food.columns)

163

Notem que podem fer **Step 5** i **Step 6** de cop, tal com segueix.

In [6]:
food.shape

(356027, 163)

La 2-tupla obtinguda conté el nombre de files i columnes del dataframe a cada component, respectivament.

### Step 7. Print the name of all the columns.

A l'**Step 4** es poden observar els noms d'algunes de les columnes de _food_. Una manera de visualitzar una llista amb tots els noms és executar

```python
food.columns.tolist().
```

Així, es visualitza la llista de noms amb un únic nom per columna. <div>
Vegem-ne un exemple amb alguns noms.

In [7]:
columns_names=food.columns.tolist()
columns_names[1:10]

['url',
 'creator',
 'created_t',
 'created_datetime',
 'last_modified_t',
 'last_modified_datetime',
 'product_name',
 'generic_name',
 'quantity']

Donat que tenim 163 columnes, s'ha emprat la funció _join()_ per llistar els noms sense salts de línea.

In [8]:
print(", ".join(columns_names))

code, url, creator, created_t, created_datetime, last_modified_t, last_modified_datetime, product_name, generic_name, quantity, packaging, packaging_tags, brands, brands_tags, categories, categories_tags, categories_en, origins, origins_tags, manufacturing_places, manufacturing_places_tags, labels, labels_tags, labels_en, emb_codes, emb_codes_tags, first_packaging_code_geo, cities, cities_tags, purchase_places, stores, countries, countries_tags, countries_en, ingredients_text, allergens, allergens_en, traces, traces_tags, traces_en, serving_size, no_nutriments, additives_n, additives, additives_tags, additives_en, ingredients_from_palm_oil_n, ingredients_from_palm_oil, ingredients_from_palm_oil_tags, ingredients_that_may_be_from_palm_oil_n, ingredients_that_may_be_from_palm_oil, ingredients_that_may_be_from_palm_oil_tags, nutrition_grade_uk, nutrition_grade_fr, pnns_groups_1, pnns_groups_2, states, states_tags, states_en, main_category, main_category_en, image_url, image_small_url, ene

### Step 8. What is the name of 105th column?

Si volem conèixer el nom de la columna 105, basta veure l'element 105 de la llista *columns_names* obtinguda a l'apartat anterior.

In [9]:
columns_names[105]

'-fructose_100g'

### Step 9. What is the type of the observations of the 105th column?

Un cop tenim el nom de la columna 105 de _food_, podem emprar _dtype_ per conèixer el tipus de valors que recull aquesta columna.

In [10]:
food['-fructose_100g'].dtype

dtype('float64')

Si no volguéssim emprar el nom <span style="color:#B43104">'-fructose_100g'</span> per indicar la columna, també podríem introduir l'índex emprant _iloc_, tal com segueix.

In [11]:
food.iloc[:,105].dtype

dtype('float64')

Així, tenim que la columna 105 està en precisió double float.  
Vegem alguns dels valors de la columna.

In [12]:
#Vector de TRUE-FALSE segons si el valor es un NaN o no.
col_105_nan=food[columns_names[105]].isnull()

#'-' per canviar els valors de col_105_nan a FALSE-TRUE
food[[columns_names[105]]][-col_105_nan][1:10]

-fructose_100g
204451            25.0
204452            29.0
204453            28.0
204454            25.0
204455            18.0
204456           100.0
204458            26.0
210114            63.9
210138            63.9

### Step 10. How is the dataset indexed?

A l'igual que hem fet a **Step 7**, per conèixer el nom de les columnes, podem emprar

```python
food.index.tolist()
```
per obtenir una llista amb els índexs de cada producte.

De nou, utilitzam _join()_  per poder visualitzar més d'un nombre per línia.  A més, com tenim 356027 observacions, veurem només els índexs corresponents a les 151 primeres.

In [13]:
#Empram  str() perque la funció només funciona per strings
print(" ".join(str(x) for x in food.index.tolist()[:151]))

0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150


Així, tenim que els productes estan indexats de 0 a 356027, però si ens fixam amb el nom de les columnes, hi ha una que correspon al codi del producte. D'aquesta manera, resulta natural indexar el dataframe _food_ per aquesta columna. <div>
Per fer-ho, basta executar les següents instruccions.

In [14]:
#Indexam pel codi
food = food.set_index(food['code'])
#Eliminam la columna codi
food = food.drop('code', axis=1)

Viasualitzem algunes files de la matriu per observar els canvis.

In [15]:
food.head()

url  \
code                                                       
3087   http://world-en.openfoodfacts.org/product/0000...   
4530   http://world-en.openfoodfacts.org/product/0000...   
4559   http://world-en.openfoodfacts.org/product/0000...   
16087  http://world-en.openfoodfacts.org/product/0000...   
16094  http://world-en.openfoodfacts.org/product/0000...   

                          creator   created_t      created_datetime  \
code                                                                  
3087   openfoodfacts-contributors  1474103866  2016-09-17T09:17:46Z   
4530              usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
4559              usda-ndb-import  1489069957  2017-03-09T14:32:37Z   
16087             usda-ndb-import  1489055731  2017-03-09T10:35:31Z   
16094             usda-ndb-import  1489055653  2017-03-09T10:34:13Z   

      last_modified_t last_modified_datetime                    product_name  \
code                                                                           
3087       1474103893   2016-09-17T09:18:13Z              Farine de blé noir   
4530       1489069957   2017-03-09T14:32:37Z  Banana Chips Sweetened (Whole)   
4559       1489069957   2017-03-09T14:32:37Z                         Peanuts   
16087      1489055731   2017-03-09T10:35:31Z          Organic Salted Nut Mix   
16094      1489055653   2017-03-09T10:34:13Z                 Organic Polenta   

      generic_name quantity packaging         ...          \
code                                          ...           
3087           NaN      1kg       NaN         ...           
4530           NaN      NaN       NaN         ...           
4559           NaN      NaN       NaN         ...           
16087          NaN      NaN       NaN         ...           
16094          NaN      NaN       NaN         ...           

      fruits-vegetables-nuts_100g fruits-vegetables-nuts-estimate_100g  \
code                                                                     
3087                          NaN                                  NaN   
4530                          NaN                                  NaN   
4559                          NaN                                  NaN   
16087                         NaN                                  NaN   
16094                         NaN                                  NaN   

      collagen-meat-protein-ratio_100g cocoa_100g chlorophyl_100g  \
code                                                                
3087                               NaN        NaN             NaN   
4530                               NaN        NaN             NaN   
4559                               NaN        NaN             NaN   
16087                              NaN        NaN             NaN   
16094                              NaN        NaN             NaN   

      carbon-footprint_100g nutrition-score-fr_100g nutrition-score-uk_100g  \
code                                                                          
3087                    NaN                     NaN                     NaN   
4530                    NaN                    14.0                    14.0   
4559                    NaN                     0.0                     0.0   
16087                   NaN                    12.0                    12.0   
16094                   NaN                     NaN                     NaN   

      glycemic-index_100g water-hardness_100g  
code                                           
3087                  NaN                 NaN  
4530                  NaN                 NaN  
4559                  NaN                 NaN  
16087                 NaN                 NaN  
16094                 NaN                 NaN  

[5 rows x 162 columns]

<br>
I ara, vegem els índexs corresponents a les primeres 151 observacions, com hem fet abans del canvi.

In [16]:
print(" ".join(str(x) for x in food.index.tolist()[:151]))

3087 4530 4559 16087 16094 16100 16117 16124 16193 16513 16612 16650 16872 16933 17497 18012 18050 18173 18197 18227 18265 18289 18319 18340 18357 18371 18388 18395 18401 18418 18449 18456 18500 18517 18524 18555 18562 18579 18630 18708 18753 18760 18869 18883 18937 18944 20114 24600 27083 27205 30038 30113 30540 30557 30564 30571 30595 30625 30649 30717 30724 31097 31141 31172 31233 31288 32070 32117 32384 32858 32896 33015 33060 33084 33268 33503 33572 33589 33626 33640 33657 33664 33688 34135 34142 34241 34265 34425 34432 34449 34456 34470 34548 34562 34623 34715 34784 34791 35071 35170 35279 35286 35293 35323 35583 35590 35613 36153 36252 36269 36276 36986 38058 38065 38072 38096 38119 38133 38317 38324 38348 38508 38522 38539 38584 38591 38607 38621 38652 38669 38690 38706 38713 38744 38973 39093 39208 39215 39259 39291 39529 39536 39543 39550 39567 39628 41355 43083 43595 45292 45308


### Step 11. What is the product name of the 19th observation?

Per accedir al nom del producte que correspon a la fila 19, basta indicar a **Phyton** que volem la columna <span style="color:#B43104">'product_name'</span> i tornar l'observació 19.

In [17]:
food['product_name'][19]

'Organic Oat Groats'